mlflow/gluon.py

```diff
@@ -0,0 +1,184 @@
+    >>> from mxnet.gluon.contrib import estimator
+    >>> from mxnet.gluon.loss import SoftmaxCrossEntropyLoss
+    >>> from mxnet.gluon.nn import HybridSequential
+    >>> from mxnet.metric import Accuracy
+    >>> import mlflow
+    >>> # Build, compile, and train your model
+    >>> net = HybridSequential()
+    >>> with net.name_scope():
+    >>> ...
+    >>> net.hybridize()
+    >>> net.collect_params().initialize()
+    >>> softmax_loss = SoftmaxCrossEntropyLoss()
+    >>> trainer = Trainer(net.collect_params())
+    >>> est = estimator.Estimator(net=net, loss=softmax_loss, metrics=Accuracy(), trainer=trainer)
+    >>> est.fit(train_data=train_data, epochs=100, val_data=validation_data)
+    >>> # Log metrics and log the model
+    >>> with mlflow.start_run() as run:
+    >>>   mlflow.gluon.log_model(net, "model")
+    """
+    Model.log(artifact_path=artifact_path, flavor=mlflow.gluon, gluon_model=gluon_model)
+
+
+@experimental
+def autolog():
+    """
+    Enable automatic logging from Gluon to MLflow.
+    Logs loss and any other metrics specified in the fit
+    function, and optimizer data as parameters. Model checkpoints
+    are logged as artifacts to a 'models' directory.
+    """
+    class __MLflowGluonCallback(TrainEnd, EpochEnd):
+        def __init__(self):
+            self.current_epoch = 0
+
+        def epoch_end(self, estimator, *args, **kwargs):
+            logs = {}
+            for metric in estimator.train_metrics:
+                metric_name, metric_val = metric.get()
+                logs[metric_name] = metric_val
+            for metric in estimator.val_metrics:
+                metric_name, metric_val = metric.get()
+                logs[metric_name] = metric_val
+            try_mlflow_log(mlflow.log_metrics, logs, step=self.current_epoch)
+            self.current_epoch += 1
+
+        def train_end(self, estimator, *args, **kwargs):
+            try_mlflow_log(mlflow.log_param, "num_layers", len(estimator.net))
+            if estimator.max_epoch is not None:
+                try_mlflow_log(mlflow.log_param, "epochs", estimator.max_epoch)
+            if estimator.max_batch is not None:
+                try_mlflow_log(mlflow.log_param, "batches", estimator.max_batch)
+            try_mlflow_log(mlflow.log_param, "optimizer_name",
+                           type(estimator.trainer.optimizer).__name__)
+            if hasattr(estimator.trainer.optimizer, "lr"):
+                try_mlflow_log(mlflow.log_param, "learning_rate",
+                               estimator.trainer.optimizer.lr)
+            if hasattr(estimator.trainer.optimizer, "epsilon"):
+                try_mlflow_log(mlflow.log_param, "epsilon",
+                               estimator.trainer.optimizer.epsilon)
+            if isinstance(estimator.net, HybridSequential):
+                try_mlflow_log(log_model, estimator.net, artifact_path="model")
+
+    @gorilla.patch(Estimator)
+    def fit(self, *args, **kwargs):
```

Comment:
As https://github.com/mlflow/mlflow/pull/2056/ does for Keras/TensorFlow, could we add logic to check if there's an active run here & if not (i.e. if we end up creating an active run within this `fit` call) add logic to end the run at the end of `fit`? It'd also be good to add a test case analogous to those in https://github.com/mlflow/mlflow/pull/2056/ for this behavior, thanks!

Reply:
I will look into this.